partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/TypeEnv.kt

```diff
@@ -35,15 +41,21 @@ internal class TypeEnv(public val schema: List<Rel.Binding>) {
     fun resolve(path: BindingPath): Rex? {
         val head: BindingName = path.steps[0]
         var tail: List<BindingName> = path.steps.drop(1)
-        var r = matchRoot(head)
+        var r = matchRoot(head, scopeIndex)
         if (r == null) {
-            r = matchStruct(head) ?: return null
+            r = matchStruct(head, scopeIndex) ?: return null
             tail = path.steps
         }
+        val op = r.op as Rex.Op.Var.Outer
+        if (op.scope == scopeIndex) {
+            r = rex(r.type, rexOpVarLocal(op.ref))
+        }
         // Convert any remaining binding names (tail) to an untyped path expression.
         return if (tail.isEmpty()) r else r.toPath(tail)
     }
 
+    private val scopeIndex: Int = this.outer.size
```

Comment:
This is interesting and deserves a comment explaining the calculation. I _almost_ think depth is a better value to use in the plan rather than the scope index because depth is relative to the VarOuter node rather than an absolute index if that makes sense.

Like a local var(0) has an implicit depth of 0, and an outer as a depth relative from 0. For an arbitrary var, what is the scope index? It depends because it's not relative to the variable .. hence the slight preference for depth

Reply:
The only thing I don't like about *depth* is having to compute the index at each execution of an *outer*. We'd need to change our implementation of `ExprVarOuter` which currently looks like:
```kotlin
    override fun eval(record: Record): PartiQLValue {
        return env[scope][reference]
    }
```

to:

```kotlin
    override fun eval(record: Record): PartiQLValue {
        val scopeIndex = env.size - depth
        return env[scopeIndex][reference]
    }
```

Reply:
You shouldn't compute this in eval. This should be computed in the compiler class, then `scope: Int` is a field of ExprVarOuter

Reply:
Both env.size and depth are known during compilation and are fixed values